crates/ruff_python_formatter/src/expression/expr_list_comp.rs

```diff
@@ -1,20 +1,41 @@
 use crate::comments::Comments;
 use crate::expression::parentheses::{
-    default_expression_needs_parentheses, NeedsParentheses, Parentheses, Parenthesize,
+    default_expression_needs_parentheses, parenthesized, NeedsParentheses, Parentheses,
+    Parenthesize,
 };
-use crate::{not_yet_implemented_custom_text, FormatNodeRule, PyFormatter};
+use crate::prelude::*;
+use crate::AsFormat;
+use crate::{FormatNodeRule, PyFormatter};
 use ruff_formatter::{write, Buffer, FormatResult};
 use rustpython_parser::ast::ExprListComp;
 
 #[derive(Default)]
 pub struct FormatExprListComp;
 
 impl FormatNodeRule<ExprListComp> for FormatExprListComp {
-    fn fmt_fields(&self, _item: &ExprListComp, f: &mut PyFormatter) -> FormatResult<()> {
+    fn fmt_fields(&self, item: &ExprListComp, f: &mut PyFormatter) -> FormatResult<()> {
+        let ExprListComp {
+            range: _,
+            elt,
+            generators,
+        } = item;
+
+        let joined = format_with(|f| {
+            f.join_with(soft_line_break_or_space())
+                .entries(generators.iter().formatted())
+                .finish()
+        });
+
         write!(
             f,
-            [not_yet_implemented_custom_text(
-                "[NOT_YET_IMPLEMENTED_generator_key for NOT_YET_IMPLEMENTED_generator_key in []]"
+            [parenthesized(
+                "[",
+                &group(&self::format_args!(
```

Comment:
Question: `parenthesized` already adds a group. Is this additional group required or should it maybe only wrap the `value` rather than the value and `joined`? This way the printer would first try to expand `joined`, and only fallback to expanding `value` if the line still exceeds the configured line width.

Reply:
i'll check, but it might also be a mistake, `parenthesized` landed after i'd written this code and i added it in when i spotted it landing

Reply:
but i should also admit i don't fully grok the rules around expanding and groups yet, and sometimes and up trying a bunch of variations on examples. your explanation in a previous comment above helps a bit though (group sequences and nesting)